### PR TITLE
zephyr: Add support for zephyr virtual filesystem

### PR DESCRIPTION
**Pull request by MaureenHelm**
Introduces the uos module to the zephyr port and adds support for
filesystem operations. Uses the zephyr virtual filesystem rather than
the micropython one since zephyr supports multiple types of filesystems
(FatFS, NFFS, LittleFS).

Limitations:
- Relative paths are not yet supported; absolute paths must be used.
- NFFS is not yet supported, but should be straightforward to add.

Example usage on the mimxrt1050_evk board with FatFS backed by an sdhc
disk:

>>> import uos
>>> fs = uos.ZfsFat()
>>> uos.mount(fs, "/SD:")
>>> f = open("/SD:/hello.txt", 'w')
>>> f.write("hello world")
>>> f.close()
>>> uos.listdir("/SD:")

Example usage on the reel_board with LittleFS backed by a flash disk:

>>> import uos
>>> fs = uos.ZfsLittle()
>>> uos.mount(fs, "/lfs")
>>> f = open("/lfs/hello.txt", 'w')
>>> f.write("hello world")
>>> f.close()
>>> uos.listdir("/lfs")

### CODE REVIEW

File: ports/zephyr/moduos.c [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+#include "zfs.h"
+#include "zfs_file.h"
+
+STATIC const mp_rom_map_elem_t uos_module_globals_table[] = {
+    { MP_ROM_QSTR(MP_QSTR___name__), MP_ROM_QSTR(MP_QSTR_uos) },
+#ifdef CONFIG_FAT_FILESYSTEM_ELM
+    { MP_ROM_QSTR(MP_QSTR_ZfsFat), MP_ROM_PTR(&zfs_fat_type) },
+#endif
+#ifdef CONFIG_FILE_SYSTEM_LITTLEFS
+    { MP_ROM_QSTR(MP_QSTR_ZfsLittle), MP_ROM_PTR(&zfs_little_type) },
```
Comment (dpgeorge) [inline ports/zephyr/moduos.c]: Does it use littlefs v1 or v2?  Would be good to rename this type to match, either `ZfsLfs1` or `ZfsLfs2` (other ports name it `VfsLfs1` and `VfsLfs2`).

**Comment by dpgeorge** [comment]
Thanks for the contribution!

> Uses the zephyr virtual filesystem rather than
> the micropython one since zephyr supports multiple types of filesystems
> (FatFS, NFFS, LittleFS).

There are multiple ways for a port to expose filesystem capabilities in MicroPython.  They are:
1. Provide block device objects (eg read/write 512 byte blocks), then use all of MicroPython's vfs system, which will then allow FAT FS, littlefs, and user defined filesystems.  It will also allow the user to create filesystems on anything, eg external SPI flash (stm32, esp8266, esp32 all do it this way).
1. Provide a VFS-capable object, then use just the core part of MicroPython's vfs system.  The VFS object is mounted under the MicroPython VFS, and moduos functions all go through the MicroPython VFS (delegating to the mounted objects).  This allows the user to still create block devices themselves and mount arbitrary filesystems (unix coverage build does it this way with vfs_posix).
1. Don't use MicroPython's VFS and just implement the uos module and functions separately (unix standard build does it this way).

What's done in this PR is approach 3, so it doesn't allow the user to plug in their own filesystem or use their own block device.

Is it possible to expose block devices in Zephyr to use approach 1?  Or go for approach 2?

File: ports/zephyr/moduos.c (hunk 1 above)
Comment (MaureenHelm) [reply ports/zephyr/moduos.c]: v2

**Comment by MaureenHelm** [comment]
> Is it possible to expose block devices in Zephyr to use approach 1? Or go for approach 2?

I experimented with approach 1 early on, but wound up going with approach 3 to allow using all the filesystems available in Zephyr. LittleFS wasn't in Micropython yet when I started this work, and NFFS still isn't.

With approach 3, most of the work adding new filesystems is in Zephyr itself, so extending them to Micropython is pretty simple. See the `#ifdef CONFIG_FAT_FILESYSTEM_ELM` and ` #ifdef CONFIG_FILE_SYSTEM_LITTLEFS` bits to see what I mean.

I'm not sure I understand approach 2, but will spend some time looking at vfs_posix.

**Comment by MaureenHelm** [comment]
Closing in favor of #5462

**Pull request closed without merging**